### PR TITLE
Fix long string doubled quote parsing — swev-id: astropy__astropy-14598

### LINKED ISSUE #71: Inconsistency in double single-quote ('') management in FITS Card
Inconsistency in double single-quote ('') management in FITS Card

### Description

The management of single-quotes in FITS cards seem correct, except sometimes when dealing with null strings, i.e. double single quotes (`''`), which sometimes are transformed into single single quotes (`'`).

E.g.:
```python
from astropy.io import fits
for n in range(60, 70):
    card1 = fits.Card('CONFIG', "x" * n + "''")
    card2 = fits.Card.fromstring(str(card1))  # Should be the same as card1
    print(n, card1.value == card2.value)
    if card1.value != card2.value:
        print(card1.value)
        print(card2.value)
```

Observed output:
```
60 True
61 True
62 True
63 True
64 True
65 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
66 True
67 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
68 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
69 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
```

If the null string `''` is included in a larger value, the issue occurs at a different position:
```python
from astropy.io import fits
for n in range(50, 70):
    card1 = fits.Card('CONFIG', "x" * n + "''" + "x"*10)
    card2 = fits.Card.fromstring(str(card1))
    print(n, len(card1.value), card1.value == card2.value)
```

Observed output:
```
50 62 True
51 63 True
52 64 True
53 65 True
54 66 True
55 67 False
56 68 False
57 69 False
58 70 False
59 71 False
60 72 False
61 73 False
62 74 False
63 75 False
64 76 True
65 77 False
66 78 True
67 79 False
68 80 False
69 81 False
```

Additional cases reported:
- Trailing quote at the end of the line is lost:
```python
fits.Card.fromstring(fits.Card("FOO", "x"*100 + "''", "comment").image).value
# => ends with just one quote
```
- A string after quotes is lost if there is a space in between:
```python
fits.Card.fromstring(fits.Card("FOO", "x"*100 + "'' aaa", "comment").image).value
# => truncates after the single quote
```

### Expected behavior

All card values should be handled properly and round-trip without losing quotes or truncating content.

### How to Reproduce

```python
from astropy.io import fits
for n in range(60, 70):
    card1 = fits.Card('CONFIG', "x" * n + "''")
    card2 = fits.Card.fromstring(str(card1))
    print(n, len(card1.value), card1.value == card2.value)
    if card1.value != card2.value:
        print(card1.value)
        print(card2.value)
```

### Versions

Linux-5.10.0-1029-oem-x86_64-with-glibc2.29
Python 3.8.10 (default, Mar 13 2023, 10:26:41)
[GCC 9.4.0]
astropy 5.2.1
Numpy 1.23.5
pyerfa 2.0.0
Scipy 1.10.0
Matplotlib 3.6.2

---

### Researcher specification

Root cause:
- In astropy/io/fits/card.py, Card._split (long-image path used for continued long string values) performs per-chunk unescaping of doubled single quotes via `replace("''", "'")`. This premature unescape corrupts semantics when doubled quotes are near chunk boundaries or followed by spaces, leading to quote loss or truncation when the reassembled value is parsed.

Proposed fix:
- Do not unescape doubled single quotes per CONTINUE chunk in Card._split. Let the final parse step handle unescaping after full reassembly. Specifically, in Card._split (long-image branch), change:
  - `value = (m.group("strg") or "").rstrip().replace("''", "'")`
  - to: `value = (m.group("strg") or "").rstrip()`
  - Keep existing handling of trailing `&` inside the quoted chunk.
- No changes to serialization logic; it already doubles embedded quotes correctly in `_format_value` and `_format_long_image`.

Optional robustness (defer unless desired):
- Consider tightening the `_strg` regex to `r"'(?P<strg>(?:[^']|'')*)'(?=$|/| )"` to ensure embedded doubled quotes are always treated as content and only a single, non-doubled `'` can terminate the string.

Tests to add (pytest):
- Round-trip for trailing doubled quotes near boundary lengths (e.g., n in 60..100) with and without comments.
- Embedded doubled quotes across boundaries within larger strings; ensure equality after round-trip.
- Reproduce and fix the two additional cases: trailing doubled quote preserved; space-separated text after doubled quotes preserved.
- Null string `''` round-trip, including when in the middle of a long continued value.

Compatibility:
- Change only affects previously buggy cases. Normal behavior is preserved; serialization unchanged.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- Preserve doubled single quotes ("''") across CONTINUE segments when parsing long string FITS Cards
- Defer unescaping until after full value reassembly; remove per-chunk unescape in `Card._split`
- Refine continuation handling: strip ampersand before recovery and avoid pulling in formatting spaces on comment cards
- Add regression tests for boundary lengths, embedded doubled quotes, null string cases, and comment interactions

## Observed Failure
Silent data corruption on round-trip of FITS Card values containing doubled quotes near 80-col boundaries:

- Trailing doubled quotes lose one quote at certain lengths:
```
65 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
67 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
68 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
69 False
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx''
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
```

- When `''` appears within a larger value, mismatches occur across a range of lengths (often around continuation boundaries):
```
55 67 False
56 68 False
57 69 False
58 70 False
59 71 False
60 72 False
61 73 False
62 74 False
63 75 False
65 77 False
67 79 False
68 80 False
69 81 False
```

- Additional reported cases:
  - Trailing quote at end of long line lost:
    - `fits.Card.fromstring(fits.Card("FOO", "x"*100 + "''", "comment").image).value` → returns a value ending with a single quote.
  - Text after doubled quotes lost if there is a space in between:
    - `fits.Card.fromstring(fits.Card("FOO", "x"*100 + "'' aaa", "comment").image).value` → truncates after the single quote.

## Reproduction Steps
Run any of the following in Python with astropy.io.fits:

```python
from astropy.io import fits

# 1) Doubled quotes at end
for n in range(60, 70):
    card1 = fits.Card('CONFIG', "x" * n + "''")
    card2 = fits.Card.fromstring(str(card1))
    assert card1.value == card2.value

# 2) Doubled quotes inside a longer string
for n in range(50, 70):
    card1 = fits.Card('CONFIG', "x" * n + "''" + "x" * 10)
    card2 = fits.Card.fromstring(str(card1))
    assert card1.value == card2.value

# 3) Additional cases
assert fits.Card.fromstring(fits.Card("FOO", "x"*100 + "''", "comment").image).value.endswith("''")
assert fits.Card.fromstring(fits.Card("FOO", "x"*100 + "'' aaa", "comment").image).value.endswith("'' aaa")
```

Expected: All assertions pass.
Actual (pre-fix): Some assertions fail at lengths near continuation boundaries.

## Stack Trace
No Python exception is raised; this is a silent parsing defect (logical data corruption). Therefore, no traceback is available. The failure is observable via round-trip mismatch as shown above.

## Implementation Notes
- Root cause: In `astropy/io/fits/card.py`, `Card._split` (long-image path) unescaped doubled quotes per CONTINUE chunk (`replace("''", "'")`), introducing stray single quotes at chunk boundaries that the final parser misinterpreted as terminators, leading to quote loss or truncation.
- Fix: Remove per-chunk unescape; let final parse unescape after full value reassembly. Ensure continuation ampersands are stripped before whitespace recovery, and skip whitespace recovery when a comment field is present.
- Serialization (`_format_value`, `_format_long_image`) unchanged; they already double embedded quotes correctly.

## Tests
- Added regression coverage in `astropy/io/fits/tests/test_header.py`:
  - Trailing doubled quotes near boundary lengths with and without comments
  - Embedded doubled quotes across boundaries within larger strings
  - Null string `''` round-trip, including within long continued values
  - Comment interaction cases (no spurious `&` or spaces)

Fixes #71

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Results
- PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 LD_LIBRARY_PATH=/nix/store/wffgswxkp55xi14jy63rjsnfvl2qvmxy-gcc-14.3.0-lib/lib:  python -m pytest -p pytest_astropy_header -p pytest_doctestplus   /workspace/astropy/.venv/lib/python3.11/site-packages/astropy/io/fits/tests/test_header.py   -k "doubled_quotes or regression_cases_with_trailing_quotes or null_string_round_trip_including_continued_value" (6 passed, 175 deselected)

File: astropy/io/fits/card.py [hunk 1]
```diff
@@ -858,20 +858,42 @@ def _split(self):
...
                 value = m.group("strg") or ""
-                value = value.rstrip().replace("''", "'")
-                if value and value[-1] == "&":
+                if value.endswith("&"):
                     value = value[:-1]
+
+                if not comment:
+                    trailing = vc[m.end("strg") : m.end(0)]
+                    if trailing:
+                        trailing_spaces = trailing.replace("'", "")
```
Comment (noa-lucent) [inline astropy/io/fits/card.py:869]: [major] The `if not comment` guard ends up treating COMM fields that are present but empty (for example a raw CONTINUE card ending in `" /"` with no text) as if there were no comment at all. When that happens `m.group('comm')` returns the empty string, we fall into this whitespace-recovery block, `trailing` becomes the comment separator (`"'/"` → `" /"` after removing the quotes), and that literal ` /` fragment gets spliced into the accumulated value. The remainder block below does the same. That breaks round-tripping for long strings that legitimately carry an empty comment field. Please change these guards to check for `comment is None` so that we skip whitespace recovery whenever a comment separator was present, even if its payload is empty.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Summary:\n- Thanks for tightening the continuation handling and regression tests.\n- I spotted one more edge case: when a CONTINUE card has a comment separator but no comment text,  yields an empty string, so the new  guards treat it as a value fragment and splice the literal  into the long-string payload.\n\nPlease change the guards to check  (and update the remainder block accordingly) so we only recover whitespace when the comment field is completely absent.

**Comment by casey-brooks** [comment]
## Test & Lint Results (update)
- PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 LD_LIBRARY_PATH=/nix/store/wffgswxkp55xi14jy63rjsnfvl2qvmxy-gcc-14.3.0-lib/lib:  python -m pytest -p pytest_astropy_header -p pytest_doctestplus   /workspace/astropy/.venv/lib/python3.11/site-packages/astropy/io/fits/tests/test_header.py   -k "doubled_quotes or regression_cases_with_trailing_quotes or null_string_round_trip_including_continued_value" (6 passed, 175 deselected)

File: astropy/io/fits/card.py (hunk 1 above)
Comment (casey-brooks) [reply astropy/io/fits/card.py:869]: Thanks for catching that edge case. I've tightened both guards to check \ so we only run the whitespace recovery when the comment field is truly absent. Targeted FITS header regressions still pass (see updated test summary above).

**Review by noa-lucent — APPROVED** [review]
Thanks for tightening the guards. Confirmed both  sites now check , so continuation whitespace recovery no longer triggers for empty-comment CONTINUE lines. Reconstructed values stay clean (no stray  or spaces), and the existing regression tests still cover the edge cases.\n\nTests not rerun in this environment.

**Review by noa-lucent — APPROVED** [review]
Thanks for tightening the guards. Confirmed both `if not comment` sites now use `comment is None`, so continuation whitespace recovery no longer triggers for empty-comment CONTINUE lines. Reconstructed values stay clean (no stray `&` or spaces), and the existing regression tests still cover the edge cases.

Tests not rerun in this environment.
